File: src/types/intersection.ts
// Types for the "Intersection" right-of-way puzzle game.

export type Direction = 'north' | 'south' | 'east' | 'west';
export type Intent = 'straight' | 'left' | 'right';
export type AgentType = 'car' | 'pedestrian' | 'cyclist' | 'emergency';
export type IntersectionType =
  | 'four-way-stop'
  | 'two-way-stop'
  | 'uncontrolled'
  | 't-intersection'
  | 'signal';

export interface Agent {
  /** Stable id, referenced by correctOrder / stepRules / collisionRules. */
  id: string;
  type: AgentType;
  /** Which arm of the intersection the agent sits on. */
  approachFrom: Direction;
  /** Where the agent is headed — drives the intent arrow. */
  intent: Intent;
  /** Optional lane refinement (bike lane for cyclists, turn lanes, etc.). */
  lane?: 'left' | 'right' | 'bike';
  /** Optional short tag shown on the agent, e.g. "Arrived first". */
  label?: string;
}

export interface IntersectionLevel {
  /** "level-1" .. "level-10" */
  id: string;
  title: string;
  description: string;
  intersectionType: IntersectionType;
  agents: Agent[];
  /** Agent ids in the legal order they may proceed. */
  correctOrder: string[];
  /** agentId -> why this agent goes when it does (shown in the recap). */
  stepRules: Record<string, string>;
  /**
   * Collision rule lookup. Key is "<tappedId>><shouldBeId>" — when the player
   * taps tappedId but shouldBeId had the right of way. A "*" wildcard key
   * provides a fallback rule for any wrong tap not otherwise listed.
   */
  collisionRules: Record<string, string>;
  /** e.g. "CVC §21801" — surfaced in the recap. */
  cvcReference?: string;
}

export interface IntersectionLevelsFile {
  levels: IntersectionLevel[];
}

export interface GameResult {
  levelId: string;
  stars: 1 | 2 | 3;
  attempts: number;
  /** ISO timestamp. */
  solvedAt: string;
}

/** Star rating from the number of failed attempts before solving. */
export function starsForAttempts(attempts: number): 1 | 2 | 3 {
  if (attempts <= 1) return 3;
  if (attempts === 2) return 2;
  return 1;
}

File: ios-native/DMVCalifornia/Core/Models/Intersection.swift
import Foundation

// MARK: - Intersection puzzle game models
// Mirrors the web app's src/types/intersection.ts so the shared
// intersection-levels.json decodes identically on both platforms.

enum IntersectionType: String, Codable {
    case fourWayStop = "four-way-stop"
    case twoWayStop = "two-way-stop"
    case uncontrolled
    case tIntersection = "t-intersection"
    case signal

    /// Human-readable label, e.g. "four way stop".
    var displayName: String {
        rawValue.replacingOccurrences(of: "-", with: " ")
    }
}

enum AgentType: String, Codable {
    case car, pedestrian, cyclist, emergency
}

enum Direction: String, Codable {
    case north, south, east, west
}

enum Intent: String, Codable {
    case straight, left, right
}

enum AgentLane: String, Codable {
    case left, right, bike
}

struct Agent: Codable, Hashable, Identifiable {
    let id: String
    let type: AgentType
    let approachFrom: Direction
    let intent: Intent
    let lane: AgentLane?
    let label: String?
}

struct IntersectionLevel: Codable, Hashable, Identifiable {
    let id: String
    let title: String
    let description: String
    let intersectionType: IntersectionType
    let agents: [Agent]
    /// Agent ids in the legal order they may proceed.
    let correctOrder: [String]
    /// agentId -> why it goes when it does (shown in the recap).
    let stepRules: [String: String]
    /// "<tappedId>><shouldBeId>" or "*" -> collision rule text.
    let collisionRules: [String: String]
    let cvcReference: String?
}

/// JSON wrapper — file shape is { "levels": [...] }.
struct IntersectionLevelsResponse: Codable {
    let levels: [IntersectionLevel]
}

/// Persisted per-level result (best run is kept).
struct IntersectionResult: Codable, Hashable {
    let levelId: String
    let stars: Int          // 1...3
    let attempts: Int       // failed attempts before solving
    let solvedAt: Date
}

/// Star rating from the number of failed attempts before solving.
func starsForAttempts(_ attempts: Int) -> Int {
    if attempts <= 1 { return 3 }
    if attempts == 2 { return 2 }
    return 1
}
